### PR TITLE
[xray] support multi-workers per process

### PR DESCRIPTION
**Pull request by raulchen**
This is the first RP to enable multi-workers per thread. This RP only include the changes on raylet backend side. Subsequent RPs will include the changes on non-raylet backend and Java worker.

Implementation details:
1. To minimize code changes and simplify implementation, right now, we only support FIXED number of workers per process. 
2. The worker process can implement concurrent workers using multi-threads or any other approaches. This is transparent to the node manager.
3. Each worker has its own connection to the node manager. Node manager looks up a worker by the connection, not by pid.

### CODE REVIEW

File: src/ray/raylet/worker_pool.cc [hunk 1]
```diff
@@ -44,20 +46,21 @@ WorkerPool::~WorkerPool() {
   pool_.clear();
```
Comment (eric-jj) [inline src/ray/raylet/worker_pool.cc]: remove all the clear operations, they are not needed.

File: src/ray/raylet/worker_pool.cc [hunk 2]
```diff
@@ -44,20 +46,21 @@ WorkerPool::~WorkerPool() {
...
   return static_cast<uint32_t>(actor_pool_.size() + pool_.size());
 }
 
-void WorkerPool::StartWorker(bool force_start) {
+void WorkerPool::StartWorkerProcess(bool force_start) {
   RAY_CHECK(!worker_command_.empty()) << "No worker command provided";
   // The first condition makes sure that we are always starting up to
   // num_cpus_ number of processes in parallel.
-  if (NumWorkersStarting() > num_cpus_ && !force_start) {
+  if (static_cast<int>(starting_worker_processes_.size()) > num_cpus_ && !force_start) {
```
Comment (eric-jj) [inline src/ray/raylet/worker_pool.cc]: should "static_cast<int>(starting_worker_processes_.size()) > num_cpus_" be >= here? When the ==, after the processing, the process count will be greater than num_cpus_.

Comment (raulchen) [reply src/ray/raylet/worker_pool.cc]: good catch

File: src/ray/raylet/worker_pool.cc [hunk 3]
```diff
@@ -67,8 +70,8 @@ void WorkerPool::StartWorker(bool force_start) {
   // Launch the process to create the worker.
   pid_t pid = fork();
   if (pid != 0) {
-    RAY_LOG(DEBUG) << "Started worker with pid " << pid;
-    started_worker_pids_.insert(pid);
+    RAY_LOG(DEBUG) << "Started worker process with pid " << pid;
+    starting_worker_processes_.emplace(std::make_pair(pid, num_workers_per_process_));
```
Comment (eric-jj) [inline src/ray/raylet/worker_pool.cc]: AddStartingWorkerProcess(pid)

**Comment by raulchen** [comment]
@eric-jj the change on java worker will be done in the next RP, because we'll need to make java worker compatible with raylet first.
For now, `num_workers_per_process` is always set to 1. Python/Java worker can still work.

**Pull request closed without merging**